internal/dispatch/remote/cluster.go

```diff
@@ -881,6 +964,46 @@ func (srst *supportedResourceSubjectTracker) updateForError(err error) {
 	}
 }
 
+type primarySleeper struct {
+	reqKey   string
+	waitTime time.Duration
+	cancel   context.CancelFunc // GUARDED_BY(lock)
+	lock     sync.Mutex
+}
+
+func (s *primarySleeper) sleep(ctx context.Context) {
+	if s.waitTime <= 0 {
+		return
+	}
+
+	s.lock.Lock()
+	ctx, s.cancel = context.WithCancel(ctx)
+	s.lock.Unlock()
+
+	hedgeWaitHistogram.WithLabelValues(s.reqKey).Observe(s.waitTime.Seconds())
+	log.Trace().Str("request-key", s.reqKey).Dur("wait-time", s.waitTime).Msg("primary dispatch waiting before running")
+
+	timer := time.NewTimer(s.waitTime)
+	defer timer.Stop()
+
+	select {
+	case <-ctx.Done():
+		log.Trace().Str("request-key", s.reqKey).Msg("primary dispatch was canceled before waiting")
+		return
+
+	case <-timer.C:
+		log.Trace().Str("request-key", s.reqKey).Msg("primary dispatch finished waiting")
+		hedgeActualWaitHistogram.WithLabelValues(s.reqKey).Observe(s.waitTime.Seconds())
```

Comment:
This should be in a defer block, because if the context is cancelled, we are not capturing the actual wait in that case.

Reply:
Exactly: if the context is canceled, it is not the actual wait time, so we do not want to log it. The actual wait time is only logged once the timer has completed